### ISSUE TITLE
`to json` colorization

### ISSUE THREAD

**Issue by NotTheDr01ds** [nushell/nushell#13444]
### Related problem

I haven't had much need for `jq` since I started using Nushell, but it still outputs a nicer JSON representation simply because it colorizes keys and types.

E.g.:

```nu
> ls | to json
# All foreground color
> ls | to json | jq
> # keys are consistently colorized and values are colorized according to type
```

### Describe the solution you'd like

* Add a `$env.config.color_config.key` that would apply to record keys as well as `to json` (and likely `to nuon`) keys.
* Apply `color_config` to values based on the type

**Comment by fdncred** [comment]
it's `to json` vs `into json` but anyway, when you do `to json` this is outputting a string. in order to do what you're talking about we'd probably have to have a nushell value for json, which we don't have now.

another option is to change the string closure in the config to do something like you're talking about. for instance, this is how i currently have string set in my theme:
```nushell
  string: {|x| if $x =~ '^#[a-fA-F\d]+' { $x } else { 'default' } }
```

**Comment by NotTheDr01ds** [comment]
> it's `to json` vs `into json`

Ah, got it right in the title, at least ;-).

> but anyway, when you do `to json` this is outputting a string. in order to do what you're talking about we'd probably have to have a nushell value for json, which we don't have now.

Bummer

> another option is to change the string closure in the config to do something like you're talking about. for instance, this is how i currently have string set in my theme:

I did consider this, but JSON nesting is a pretty hairy exercise in regex, IIRC.

**Comment by sholderbach** [comment]
Sounds like a job for `PipelineMetadata` :) 

i.e. when converting with `to json` attach a JSON pipeline metadata to the string and `table` can then decide to color, but everything else would be unaffected.

**Comment by KAAtheWiseGit** [comment]
This can also be done with `bat`:

```Nushell
~> alias colorize = bat -l json --style plain --paging never
```
![image](https://github.com/user-attachments/assets/80a598f7-7971-45c0-beb0-642f231d883e)

**Comment by Bahex** [comment]
Since 0.98.0 (#13785), it's possible to have something like this in `$env.config.hooks.display_output`
```nu
metadata access {|meta| match $meta.content_type? {
	"application/x-nuscript" | "application/x-nuon" | "text/x-nushell" => { nu-highlight },
	"application/json" => { bat --language=json --style=plain --paging=never },
	_ => {},
}}
| if (term size).columns >= 100 { table -e } else { table }
```

**Comment by fdncred** [comment]
@Bahex That's pretty creative. Thanks! I just can't get `open -r somefile.nu` to highlight. I'm not exactly sure why though.

**Comment by KAAtheWiseGit** [comment]
Since it depends on the pipeline metadata, I don't think opening files would work.  You'd need to add a case for checking `source`:

```Nushell
if $meta.source ends-with ".nu" {
    nu-highlight
} else {
    match meta.content_type? {
         ...
    }
}
```

**Comment by fdncred** [comment]
I changed `open` to assign it. I should've mentioned that.
changes in open.rs
```rust
let content_type = path
    .extension()
    .map(|ext| ext.to_string_lossy().to_string())
    .and_then(|ref s| detect_content_type(s));

...
fn detect_content_type(extension: &str) -> Option<String> {
    // This will allow the overriding of metadata to be consistent with
    // the content type
    match extension {
        // Per RFC-9512, application/yaml should be used
        "yaml" | "yml" => Some("application/yaml".to_string()),
        "nu" => Some("application/x-nuscript".to_string()),
        "json" | "jsonl" | "ndjson" => Some("application/json".to_string()),
        "nuon" => Some("application/x-nuon".to_string()),
        _ => mime_guess::from_ext(extension)
            .first()
            .map(|mime| mime.to_string()),
    }
}
```
changes in byte_stream.rs
```nushell
    fn is_external(&self) -> bool {
        matches!(self, ByteStreamSource::Child(..))
    }
```

**Comment by KAAtheWiseGit** [comment]
Nice!

One thing though is that both jsonl and ndjson recommend a different application type: [*application/jsonl*][jsonl] and [*application/x-ndjson*][ndjson].  I don't know what's best here, since I doubt these types are widespread.

[jsonl]: https://jsonlines.org/#:~:text=MIME%20type%20may%20be%20application/jsonl
[ndjson]: https://github.com/ndjson/ndjson-spec#:~:text=The%20MediaType%20[RFC6838]%20for%20Newline%20Delimited%20JSON%20SHOULD%20be%20application/x-ndjson

**Comment by fdncred** [comment]
ok, that's fine but it doesn't explain why `open script.nu` doesn't get highlighted.

**Comment by KAAtheWiseGit** [comment]
Ah, I misunderstood.  You use the `main` build, right?  Because application type doesn't get set for me

**Comment by KAAtheWiseGit** [comment]
Actually, neither does `yml`, despite a check for it being present in the 0.99.0 version.  There might be a bug

**Comment by fdncred** [comment]
> Ah, I misunderstood. You use the `main` build, right? Because application type doesn't get set for me

ya, i used main with the code changes i posted above

**Comment by KAAtheWiseGit** [comment]
Got it.  `display_output` isn't evaluated for byte streams:

```Rust
# crates/nu-cli/src/util.rs#L286

    if let PipelineData::ByteStream(..) = pipeline {
        pipeline.print(engine_state, stack, false, false)
    } else if let Some(hook) = engine_state.get_config().hooks.display_output.clone() {
        let pipeline = eval_hook(
            engine_state,
            stack,
            Some(pipeline),
            vec![],
            &hook,
            "display_output",
        )?;
        pipeline.print(engine_state, stack, false, false)
    } else {
        pipeline.print(engine_state, stack, true, false)
    }?;
```

That's why the hook doesn't work.  It never gets evaluated

**Comment by fdncred** [comment]
good sleuthing, thanks!t

**Comment by KAAtheWiseGit** [comment]
It appears it was done this way from the beginning, when the hook was added in beec658872.  That was two years ago and back then byte streams were external streams, so the circumstances might've changed.

**Comment by fdncred** [comment]
Thanks. I added https://github.com/nushell/nushell/pull/14141 to try and help with this. Not sure it's exactly right but it seems to work so far.

**Comment by Bahex** [comment]
This seems to be solved. We can probably add a cookbook entry and close this issue.

For anyone who finds this issue before the cookbook entry, here's an example.
```nushell
let mime_to_lang = {
    application/json: json,
    application/xml: xml,
    application/yaml: yaml,
    text/csv: csv,
    text/tab-separated-values: tsv,
    text/x-toml: toml,
    text/markdown: markdown,
}

$env.config.hooks.display_output = {
    metadata access {|meta| match $meta.content_type? {
        null => {}
        "application/x-nuscript" | "application/x-nuon" | "text/x-nushell" => { nu-highlight },
        $mimetype if $mimetype in $mime_to_lang => { ^bat -Ppf --language=($mime_to_lang | get $mimetype) },
        _ => {},
    }}
    | if (term size).columns >= 100 { table -e } else { table }
}

```

**Comment by NotTheDr01ds** [comment]
I think it's a great workaround, but I'd love to keep this open for a potential native implementation

**Comment by cptpiepmatz** [comment]
I don't think we should include highlighting for any language other than nushell and nuon. Other tools and plugins can do this very well.
